### PR TITLE
feat: comply with new Infura API

### PR DESCRIPTION
**Pull request by pgebal**
References https://github.com/omgnetwork/elixir-omg/issues/1752
Changes:
- sets `from` argument in `eth_call` requests
- removes flag `halt_exit_status` from dialyzer because it's no longer a valid argument

Does not address Infura API changes in error response format because:
- it is handled well in `ethereumex`
- `eth_call` error response format changes only on Kovan.
- we do not call `eth_estimateGas` 

When playing with current Infura API I could provide any address as `from` argument to `eth_call`.
We could probably do well with hardcoding any address as `from`, but to keep it nice I decided to use authority address for the calls.

### CODE REVIEW

**Comment by pgebal** [comment]
I guess some work in infra repos has to be done when introducing a new env var. I'll talk to infra guys before merging this.

File: apps/omg_eth/lib/omg_eth/release_tasks/set_ethereum_client.ex [hunk 1]
```diff
@@ -66,6 +76,16 @@ defmodule OMG.Eth.ReleaseTasks.SetEthereumClient do
   defp validate_string(value, _default) when is_binary(value), do: value
   defp validate_string(_, default), do: default
 
+  defp validate_address("0x" <> hex = address, _default) do
+    case Base.decode16(hex, case: :mixed) do
+      {:ok, <<_::binary-size(20)>>} -> address
+      _ -> exit("Eth call from address must be set.")
```
Comment (ayrat555) [inline apps/omg_eth/lib/omg_eth/release_tasks/set_ethereum_client.ex:82]:
```suggestion
      _ -> exit("Eth call from address is invalid.")
```

File: apps/omg_eth/test/fixtures.exs [hunk 2]
```diff
@@ -72,7 +72,8 @@ defmodule OMG.Eth.Fixtures do
 
   defp call_contract(contract, signature, args, return_types) do
     data = ABI.encode(signature, args)
-    {:ok, return} = Ethereumex.HttpClient.eth_call(%{to: contract, data: Encoding.to_hex(data)})
+    from = Application.fetch_env!(:omg_eth, :eth_call_from_address)
```
Comment (ayrat555) [inline apps/omg_eth/test/fixtures.exs:75]: you call everywhere `Application.fetch_env!(:omg_eth, :eth_call_from_address)`

Can you create a wrapper, for example, in `OMG.Configuration` for this function?

**Review by ayrat555 — APPROVED** [review]
a couple of style comments. lgtm

**Comment by InoMurko** [comment]
The only address the childchain has access to is the authority address. Shouldn't that account be used as the `from` parameter?

File: apps/omg_conformance/test/support/conformance/merkle_proofs.ex [hunk 3]
```diff
@@ -58,7 +58,11 @@ defmodule Support.Conformance.MerkleProofs do
   defp call_contract(contract, signature, args, return_types) do
     data = ABI.encode(signature, args)
 
-    {:ok, return} = Ethereumex.HttpClient.eth_call(%{to: Encoding.to_hex(contract), data: Encoding.to_hex(data)})
+    from = Application.fetch_env!(:omg_eth, :eth_call_from_address)
+
+    {:ok, return} =
+      Ethereumex.HttpClient.eth_call(%{from: from, to: Encoding.to_hex(contract), data: Encoding.to_hex(data)})
+
```
Comment (InoMurko) [inline apps/omg_conformance/test/support/conformance/merkle_proofs.ex:65]: this seems a bit unstructured, you're accessing the configuration from too deep. perhaps all that's needed to make this call should be sent as a function argument

File: apps/omg_conformance/test/support/conformance/signatures_hashes.ex [hunk 4]
```diff
@@ -132,8 +132,11 @@ defmodule Support.Conformance.SignaturesHashes do
 
   defp call_contract(contract, signature, args, return_types) do
     data = ABI.encode(signature, args)
+    from = Application.fetch_env!(:omg_eth, :eth_call_from_address)
```
Comment (InoMurko) [inline apps/omg_conformance/test/support/conformance/signatures_hashes.ex:135]: this seems a bit unstructured, you're accessing the configuration from too deep. perhaps all that's needed to make this call should be sent as a function argument

File: apps/omg_eth/lib/omg_eth/release_tasks/set_ethereum_client.ex [hunk 5]
```diff
@@ -52,6 +53,15 @@ defmodule OMG.Eth.ReleaseTasks.SetEthereumClient do
     rpc_client_type
   end
 
+  defp get_eth_call_from_address() do
+    eth_call_from_address =
+      validate_address(get_env("ETH_CALL_FROM_ADDRESS"), Application.get_env(@app, :eth_call_from_address))
```
Comment (InoMurko) [inline apps/omg_eth/lib/omg_eth/release_tasks/set_ethereum_client.ex:58]: yeah, so... shouldn't this simply be the authority address? why does this have to be a different address?

File: apps/omg_eth/lib/omg_eth/root_chain/rpc.ex [hunk 6]
```diff
@@ -20,7 +20,8 @@ defmodule OMG.Eth.RootChain.Rpc do
 
   def call_contract(client \\ Ethereumex.HttpClient, contract, signature, args) do
     data = signature |> ABI.encode(args) |> Encoding.to_hex()
-    client.eth_call(%{to: contract, data: data})
+    from = Application.fetch_env!(:omg_eth, :eth_call_from_address)
```
Comment (InoMurko) [inline apps/omg_eth/lib/omg_eth/root_chain/rpc.ex:23]: this feels very ugly, accessing the configuration way to deep in the stack, where all calls coming from?

File: apps/omg_eth/lib/omg_eth/release_tasks/set_ethereum_client.ex (hunk 5 above)
Comment (pgebal) [reply apps/omg_eth/lib/omg_eth/release_tasks/set_ethereum_client.ex:58]: > yeah, so... shouldn't this simply be the authority address? why does this have to be a different address?

I was thinking about that. I think we do not have access to authority address in watcher and I wanted to keep things consistent. I think it's the old mentality of having child-chain and watcher together. But you are right. It doesn't make sense, we can use authority address.

Comment (InoMurko) [reply apps/omg_eth/lib/omg_eth/release_tasks/set_ethereum_client.ex:58]: or simply use the plasma_framework contract address everywhere!

File: apps/omg_conformance/test/support/conformance/merkle_proofs.ex (hunk 3 above)
Comment (pgebal) [reply apps/omg_conformance/test/support/conformance/merkle_proofs.ex]: This argument is irrelevant, we could easily hardcode anything into it. I would rather leave it here than pretend it's something meaningful that has to be passed from higher levels.

File: apps/omg_conformance/test/support/conformance/signatures_hashes.ex (hunk 4 above)
Comment (pgebal) [reply apps/omg_conformance/test/support/conformance/signatures_hashes.ex]: This argument is irrelevant, we could easily hardcode anything into it. I would rather leave it here than pretend it's something meaningful that has to be passed from higher levels.

File: apps/omg_eth/lib/omg_eth/root_chain/rpc.ex (hunk 6 above)
Comment (pgebal) [reply apps/omg_eth/lib/omg_eth/root_chain/rpc.ex]: In this case I decided to move it a level higher, where we access `OMG.Eth.Configuration`.

File: priv/perf/apps/load_test/lib/child_chain/exit.ex [hunk 7]
```diff
@@ -84,6 +84,7 @@ defmodule LoadTest.ChildChain.Exit do
 
     {:ok, receipt_enc} =
       Ethereumex.HttpClient.eth_call(%{
+        from: Application.fetch_env!(:load_test, :eth_call_from_address),
```
Comment (InoMurko) [inline priv/perf/apps/load_test/lib/child_chain/exit.ex:87]: why is this a special case? why not use authority address, or plasma framework contract address?

File: apps/omg_conformance/test/support/conformance/signatures_hashes.ex [hunk 8]
```diff
@@ -132,8 +133,11 @@ defmodule Support.Conformance.SignaturesHashes do
 
   defp call_contract(contract, signature, args, return_types) do
     data = ABI.encode(signature, args)
+    from = Configuration.authority_address()
+
+    {:ok, return} =
+      Ethereumex.HttpClient.eth_call(%{from: from, to: Encoding.to_hex(contract), data: Encoding.to_hex(data)})
```
Comment (InoMurko) [inline apps/omg_conformance/test/support/conformance/signatures_hashes.ex:139]: you already have a contract address `to: Encoding.to_hex(contract)`. Could you not use the same address?

File: apps/omg_conformance/test/support/conformance/merkle_proofs.ex [hunk 9]
```diff
@@ -58,7 +59,13 @@ defmodule Support.Conformance.MerkleProofs do
   defp call_contract(contract, signature, args, return_types) do
     data = ABI.encode(signature, args)
 
-    {:ok, return} = Ethereumex.HttpClient.eth_call(%{to: Encoding.to_hex(contract), data: Encoding.to_hex(data)})
+    {:ok, return} =
+      Ethereumex.HttpClient.eth_call(%{
+        from: Configuration.authority_address(),
+        to: Encoding.to_hex(contract),
+        data: Encoding.to_hex(data)
+      })
```
Comment (InoMurko) [inline apps/omg_conformance/test/support/conformance/merkle_proofs.ex:67]:
```suggestion
    {:ok, return} =
      Ethereumex.HttpClient.eth_call(%{
        from: Encoding.to_hex(contract),
        to: Encoding.to_hex(contract),
        data: Encoding.to_hex(data)
      })
```
```suggestion
    {:ok, return} =
      Ethereumex.HttpClient.eth_call(%{
        from: Configuration.authority_address(),
        to: Encoding.to_hex(contract),
        data: Encoding.to_hex(data)
      })
```

File: apps/omg_conformance/test/support/conformance/signatures_hashes.ex (hunk 8 above)
Comment (InoMurko) [reply apps/omg_conformance/test/support/conformance/signatures_hashes.ex:139]:
```suggestion
      Ethereumex.HttpClient.eth_call(%{from: Encoding.to_hex(contract), to: Encoding.to_hex(contract), data: Encoding.to_hex(data)})
```
```suggestion
      Ethereumex.HttpClient.eth_call(%{from: from, to: Encoding.to_hex(contract), data: Encoding.to_hex(data)})
```

File: apps/omg_conformance/test/support/conformance/signatures_hashes.ex [hunk 10]
```diff
@@ -132,8 +133,11 @@ defmodule Support.Conformance.SignaturesHashes do
 
   defp call_contract(contract, signature, args, return_types) do
     data = ABI.encode(signature, args)
+    from = Configuration.authority_address()
```
Comment (InoMurko) [inline apps/omg_conformance/test/support/conformance/signatures_hashes.ex:136]:
```suggestion

```
```suggestion
    from = Configuration.authority_address()
```

File: apps/omg_eth/lib/omg_eth/root_chain/rpc.ex [hunk 11]
```diff
@@ -18,9 +18,9 @@ defmodule OMG.Eth.RootChain.Rpc do
   require Logger
   alias OMG.Eth.Encoding
 
-  def call_contract(client \\ Ethereumex.HttpClient, contract, signature, args) do
+  def call_contract(client \\ Ethereumex.HttpClient, contract, signature, args, from) do
     data = signature |> ABI.encode(args) |> Encoding.to_hex()
-    client.eth_call(%{to: contract, data: data})
+    client.eth_call(%{from: from, to: contract, data: data})
```
Comment (InoMurko) [inline apps/omg_eth/lib/omg_eth/root_chain/rpc.ex:23]:
```suggestion
    client.eth_call(%{from: contract, to: contract, data: data})
```
```suggestion
    client.eth_call(%{from: from, to: contract, data: data})
```

File: apps/omg_eth/lib/omg_eth/root_chain/rpc.ex [hunk 12]
```diff
@@ -18,9 +18,9 @@ defmodule OMG.Eth.RootChain.Rpc do
   require Logger
   alias OMG.Eth.Encoding
 
-  def call_contract(client \\ Ethereumex.HttpClient, contract, signature, args) do
+  def call_contract(client \\ Ethereumex.HttpClient, contract, signature, args, from) do
```
Comment (InoMurko) [inline apps/omg_eth/lib/omg_eth/root_chain/rpc.ex:21]:
```suggestion
  def call_contract(client \\ Ethereumex.HttpClient, contract, signature, args) do
```

File: priv/perf/apps/load_test/lib/child_chain/exit.ex (hunk 7 above)
Comment (InoMurko) [inline priv/perf/apps/load_test/lib/child_chain/exit.ex:87]:
```suggestion
        from: Application.fetch_env!(:load_test, :contract_address_plasma_framework),
```

File: priv/perf/apps/load_test/lib/ethereum/ethereum.ex [hunk 13]
```diff
@@ -136,7 +136,12 @@ defmodule LoadTest.Ethereum do
   defp get_external_data(address, signature, params) do
     data = signature |> ABI.encode(params) |> Encoding.to_hex()
 
-    {:ok, data} = Ethereumex.HttpClient.eth_call(%{to: address, data: data})
+    {:ok, data} =
+      Ethereumex.HttpClient.eth_call(%{
+        from: Application.fetch_env!(:load_test, :eth_call_from_address),
```
Comment (InoMurko) [inline priv/perf/apps/load_test/lib/ethereum/ethereum.ex:141]:
```suggestion
        from: address,
```

File: priv/perf/config/config.exs [hunk 14]
```diff
@@ -27,7 +27,8 @@ config :load_test,
   faucet_deposit_amount: trunc(:math.pow(10, 14)),
   fee_amount: 1,
   deposit_finality_margin: 10,
-  gas_price: 2_000_000_000
+  gas_price: 2_000_000_000,
+  eth_call_from_address: "0x0000000000000000000000000000000000000001"
```
Comment (InoMurko) [inline priv/perf/config/config.exs:31]:
```suggestion
  
```

**Review by InoMurko — CHANGES_REQUESTED** [review]

File: priv/perf/apps/load_test/lib/child_chain/exit.ex (hunk 7 above)
Comment (pgebal) [reply priv/perf/apps/load_test/lib/child_chain/exit.ex:87]: This is not under umbrella app. We do not have authority address here. We could use contract address to make it simple. That's what I'll do.

**Review by InoMurko — APPROVED** [review]
excellent!

**Pull request merged**